src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -31,11 +31,110 @@ public struct CallGraphNode
             public QsNullable<ImmutableArray<ResolvedType>> TypeArgs;
         }
 
+        // TODO: 
+        // This is the method that should be invoked to verify cycles of interest,
+        // i.e. where each callable in the cycle is type parametrized.
+        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined. 
+        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        {
+            var parent = rootNode.CallableName;
+            var validResolution = TryCombineTypeResolutions(parent, out var combined, edges.Select(edge => edge.ParamResolutions).ToArray());
+            var resolvedToConcrete = combined.Values.All(res => !(res.Resolution is ResolvedTypeKind.TypeParameter tp) || tp.Item.Origin.Equals(parent));
+            return validResolution && resolvedToConcrete;
+            //var isClosedCycle = validCycle && combined.Values.Any(res => res.Resolution is ResolvedTypeKind.TypeParameter tp && EqualsParent(tp.Item.Origin));
+            // TODO: check that monomorphization correctly processes closed cycles - meaning add a test...
+        }
+
+        /// <summary>
+        /// Combines subsequent concretions as part of a nested expression, or concretions as part of a cycle in the call graph,
+        /// into a single dictionary containing the resolution for the type parameters of the specified parent callable. 
+        /// The given resolutions are expected to be ordered starting with the dictionary containing the initial mapping for the 
+        /// type parameters of the specified parent callable (the "innermost resolutions"). This mapping may potentially be to 
+        /// type parameters of other callables, which are then further concretized by subsequent resolutions. 
+        /// Returns the constructed dictionary as out parameter. Returns true if the combination of the given resolutions is valid, 
+        /// i.e. if there are no conflicting resolutions and type parameters of the parent callables are uniquely resolved 
+        /// to either a concrete type, a type parameter of another callable, or themselves.
+        /// Throws an ArgumentNullException if the given parent is null. 
+        /// Throws an ArgumentException if the given resolutions imply that type parameters from multiple callables are 
+        /// simultaneously treated as concrete types. 
+        /// NOTE: This routine prioritizes the verifications to ensure the correctness of the resolution over performance.  
+        /// </summary>
+        public static bool TryCombineTypeResolutions
+            (QsQualifiedName parent, out ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType> combined,
```

Comment:
I'm having a difficult time understanding what the role of `parent` is. What is this callable in relation to the given resolutions?

Reply:
A code example would go a long way here.

Reply:
If we wanted to see if we can be more clever than having to check the resolution for each node in a cycle, then parent would get eliminated. I'd rather not start to have code example in documenting comments - the tests give these examples.